esdm/esdm_openssl.c

```diff
@@ -164,56 +167,167 @@ struct esdm_openssl_drng_state {
+
+out:
+	if (ret) {
+		OPENSSL_free(*entropy_buf);
+		OPENSSL_free(*nonce_buf);
+		*entropy_buf = NULL;
+		*nonce_buf = NULL;
+	}
+
+	esdm_openssl_hash_dealloc(hash);
+	return ret;
+}
+
 static int esdm_openssl_drbg_seed(void *drng, const uint8_t *inbuf,
 				  size_t inbuflen)
 {
-	OSSL_PARAM params[3];
+	OSSL_PARAM params[3] = { 0 };
 	struct esdm_openssl_drng_state *state = drng;
+	uint8_t *entropy_buf = NULL;
+	uint8_t *nonce_buf = NULL;
+	bool free_buffers = true;
+	int ret = 0;
+
+	if (esdm_openssl_drbg_derive_seed_nonce(inbuf, inbuflen, &entropy_buf,
+						!state->seeded,
+						&nonce_buf) != 0) {
+		esdm_logger(LOGGER_ERR, LOGGER_C_MD,
+			    "Failed to derive seed material: %s\n",
+			    ERR_error_string(ERR_get_error(), NULL));
+		ret = -EFAULT;
+		goto out;
+	};
 
 	if (!state->seeded) {
 		params[0] = OSSL_PARAM_construct_octet_string(
-			OSSL_RAND_PARAM_TEST_ENTROPY, (void *)inbuf,
-			inbuflen / 2);
+			OSSL_RAND_PARAM_TEST_ENTROPY, (void *)entropy_buf,
+			(size_t)EVP_MD_size(ESDM_OPENSSL_HASH));
 		params[1] = OSSL_PARAM_construct_octet_string(
-			OSSL_RAND_PARAM_TEST_NONCE,
-			(void *)(inbuf + inbuflen / 2), inbuflen / 2);
+			OSSL_RAND_PARAM_TEST_NONCE, (void *)nonce_buf,
+			(size_t)EVP_MD_size(ESDM_OPENSSL_HASH));
 		params[2] = OSSL_PARAM_construct_end();
+
 		if (!EVP_RAND_instantiate(state->seed_source, state->strength,
 					  0, NULL, 0, params)) {
 			esdm_logger(LOGGER_ERR, LOGGER_C_MD,
 				    "Failed to instantiate seed source: %s\n",
 				    ERR_error_string(ERR_get_error(), NULL));
-			return -EFAULT;
+			ret = -EFAULT;
+			goto out;
 		}
+		free_buffers = false;
 
 		if (!EVP_RAND_instantiate(state->drbg, state->strength, 0,
 					  (unsigned char *)"", 0, NULL)) {
 			esdm_logger(LOGGER_ERR, LOGGER_C_MD,
 				    "Failed to instantiate DRBG: %s\n",
 				    ERR_error_string(ERR_get_error(), NULL));
-			return -EFAULT;
+			ret = -EFAULT;
+			goto out;
 		}
 
 		state->seeded = 1;
 	} else {
 		params[0] = OSSL_PARAM_construct_octet_string(
-			OSSL_RAND_PARAM_TEST_ENTROPY, (void *)inbuf, inbuflen);
+			OSSL_RAND_PARAM_TEST_ENTROPY, (void *)entropy_buf,
+			(size_t)EVP_MD_size(ESDM_OPENSSL_HASH));
 		params[1] = OSSL_PARAM_construct_end();
+
 		if (!EVP_RAND_CTX_set_params(state->seed_source, params)) {
 			esdm_logger(LOGGER_ERR, LOGGER_C_MD,
 				    "Failed to reseed seed source: %s\n",
 				    ERR_error_string(ERR_get_error(), NULL));
-			return -EFAULT;
+			ret = -EFAULT;
+			goto out;
 		}
+		free_buffers = false;
 
 		if (!EVP_RAND_reseed(state->drbg, 0, NULL, 0, NULL, 0)) {
 			esdm_logger(LOGGER_ERR, LOGGER_C_MD,
 				    "Failed to reseed DRBG\n");
-			return -EFAULT;
+			ret = -EFAULT;
+			goto out;
 		}
 	}
 
-	return 0;
+out:
+	/*
+	 * both buffers are under memory management of OpenSSL now, only zeroize here
+	 * TEST-RAND does not zeroize itself. As we completely control seeding, zeroizing should
+	 * be safe.
+	 */
+	if (entropy_buf != NULL) {
+		memset_secure(entropy_buf, 0,
```

Comment:
Are you sure about the memset here? You mention that the buffer is in OpenSSL's hand now which may use it after this code completes. Doing a memset(0) here would imply that there is a possibility that OpenSSL is seeded with a NULL-buffer, no?

Reply:
I can't completely rule this out and better remove this line again. You're right.